File: php/src/Advertisements/User/Domain/Exceptions/UserNotFoundException.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisements\User\Domain\Exceptions;

use Demo\App\Common\Domain\DomainException;

final class UserNotFoundException extends DomainException
{
    private const string ADMIN_USER_NOT_FOUND_MESSAGE = 'Admin user not found';
    private const string MEMBER_USER_NOT_FOUND_MESSAGE = 'Member user not found';

    private function __construct(string $message)
    {
        parent::__construct($message);
    }

    public static function asAdmin(): self
    {
        return new self(self::ADMIN_USER_NOT_FOUND_MESSAGE);
    }

    public static function asMember(): self
    {
        return new self(self::MEMBER_USER_NOT_FOUND_MESSAGE);
    }

    public function message(): string
    {
        return $this->getMessage();
    }
}

File: kotlin/src/main/kotlin/advertisements/user/domain/exceptions/UserNotFoundException.kt
package advertisements.user.domain.exceptions

import common.domain.DomainException

class UserNotFoundException private constructor(message: String) : DomainException(message) {

    companion object {
        private const val ADMIN_USER_NOT_FOUND_MESSAGE = "Admin user not found"
        private const val MEMBER_USER_NOT_FOUND_MESSAGE = "Member user not found"

        fun asAdmin(): UserNotFoundException {
            return UserNotFoundException(ADMIN_USER_NOT_FOUND_MESSAGE)
        }

        fun asMember(): UserNotFoundException {
            return UserNotFoundException(MEMBER_USER_NOT_FOUND_MESSAGE)
        }
    }

    fun message(): String {
        return this.message ?: ""
    }
}
